src/main/java/org/quiltmc/loader/api/config/Config.java

```diff
@@ -0,0 +1,420 @@
+	 *         <li>An object whose class follows these rules</li>
+	 *     </ul>
+	 * </ul>
+	 *
+	 * @param modId the mod owning the resulting config file
+	 * @param id the config's id
+	 * @param before a {@link Creator} that can be used to configure the resulting config further
+	 * @param configCreatorClass a class as described above
+	 * @return a {@link ConfigWrapper<C>}
+	 */
+	static <C> ConfigWrapper<C> create(String modId, String id, Creator before, Class<C> configCreatorClass) {
+		return ConfigImpl.create(modId, id, Paths.get(""), before, configCreatorClass, builder -> {});
+	}
+
+	/**
+	 * Creates and registers a config with fields derived from the fields of the passed class
+	 *
+	 * <p>The passed class should have the following characteristics:
+	 * <ul>
+	 *     <li>Has a public no-argument constructor</li>
+	 *     <li>Each non-static field should be final and be one of the following types:</li>
+	 *     <ul>
+	 *         <li>A basic type (int, long, float, double, boolean, or String)</li>
+	 *         <li>A complex type (a {@link ValueList} or {@link ValueMap} of basic or complex types)</li>
+	 *         <li>An object whose class follows these rules</li>
+	 *     </ul>
+	 * </ul>
+	 *
+	 * @param modId the mod owning the resulting config file
+	 * @param id the config's id
+	 * @param configCreatorClass a class as described above
+	 * @param after a {@link Creator} that can be used to configure the resulting config further
+	 * @return a {@link ConfigWrapper<C>}
+	 */
+	static <C> ConfigWrapper<C> create(String modId, String id, Class<C> configCreatorClass, Creator after) {
+		return ConfigImpl.create(modId, id, Paths.get(""), builder -> {}, configCreatorClass, after);
+	}
+
+	/**
+	 * Creates and registers a config with fields derived from the fields of the passed class
+	 *
+	 * <p>The passed class should have the following characteristics:
+	 * <ul>
+	 *     <li>Has a public no-argument constructor</li>
+	 *     <li>Each non-static field should be final and be one of the following types:</li>
+	 *     <ul>
+	 *         <li>A basic type (int, long, float, double, boolean, or String)</li>
+	 *         <li>A complex type (a {@link ValueList} or {@link ValueMap} of basic or complex types)</li>
+	 *         <li>An object whose class follows these rules</li>
+	 *     </ul>
+	 * </ul>
+	 *
+	 * @param modId the mod owning the resulting config file
+	 * @param id the config's id
+	 * @param configCreatorClass a class as described above
+	 * @return a {@link ConfigWrapper<C>}
+	 */
+	static <C> ConfigWrapper<C> create(String modId, String id, Class<C> configCreatorClass) {
+		return ConfigImpl.create(modId, id, Paths.get(""), builder -> {}, configCreatorClass, builder -> {});
+	}
+
+	interface UpdateCallback {
+		void onUpdate(Config config);
+	}
+
+	interface Creator {
+		void create(Builder builder);
+	}
+
+	@ApiStatus.NonExtendable
+	interface Builder {
```

Comment:
It could extend `SectionBuilder` to adhere to OOP best practices, but I'd probably leave the methods declared explicitly as well for visibility purposes, at which point it wouldn't make much of a difference.

Reply:
The difference is that someone could write generic code over it.

Reply:
There's also no self type in Java, so the return types of the methods would be wrong unless I did some very ugly generics or widening of the method return types.

Reply:
Do interfaces not support bridge methods?

Reply:
I had a type mismatch, I can totally make this happen.